src/type-hierarchy.ts

```diff
@@ -0,0 +1,218 @@
+import * as path from 'path';
+import * as vscode from 'vscode';
+import * as vscodelc from 'vscode-languageclient';
+
+export namespace TypeHierarchyDirection {
+export const Children = 0;
+export const Parents = 1;
+export const Both = 2;
+}
+
+type TypeHierarchyDirection = 0|1|2;
+
+interface TypeHierarchyParams extends vscodelc.TextDocumentPositionParams {
+  resolve?: number;
+  direction: TypeHierarchyDirection;
+}
+
+interface TypeHierarchyItem {
+  name: string;
+  detail?: string;
+  kind: vscodelc.SymbolKind;
+  deprecated?: boolean;
+  uri: string;
+  range: vscodelc.Range;
+  selectionRange: vscodelc.Range;
+  parents?: TypeHierarchyItem[];
+  children?: TypeHierarchyItem[];
+  data?: any;
+}
+
+export namespace TypeHierarchyRequest {
+export const type =
+    new vscodelc
+        .RequestType<TypeHierarchyParams, TypeHierarchyItem|null, void, void>(
+            'textDocument/typeHierarchy');
+}
+
+interface ResolveTypeHierarchyItemParams {
+  item: TypeHierarchyItem;
+  resolve: number;
+  direction: TypeHierarchyDirection;
+}
+
+export namespace ResolveTypeHierarchyRequest {
+export const type =
+    new vscodelc.RequestType<ResolveTypeHierarchyItemParams,
+                             TypeHierarchyItem|null, void, void>(
+        'typeHierarchy/resolve');
+}
+
+class TypeHierarchyTreeItem extends vscode.TreeItem {
+  constructor(item: TypeHierarchyItem) {
+    super(item.name);
+    if (item.children && item.children.length > 0) {
+      this.collapsibleState = vscode.TreeItemCollapsibleState.Expanded;
+    } else {
+      this.collapsibleState = vscode.TreeItemCollapsibleState.Collapsed;
```

Comment:
don't show the disclosure widget if we know there are no children.
i.e. children != null && children.length == 0 ==> TreeItemCollapsibleState.None